primap2/pm2io/_data_reading.py

```diff
@@ -1244,17 +1310,150 @@ def rename_columns(
     return attrs
 
 
+_special_codes = {
+    "C": np.nan,
+    "NaN": np.nan,
+    "nan": np.nan,
+    "-": 0,
+    "NE0": np.nan,
+    "": np.nan,
+}
+
+
+def is_float(to_test: Any) -> bool:
+    try:
+        float(to_test)
+        return True
+    except ValueError:
+        return False
+
+
+def find_str_values_in_data(
+    data: pd.DataFrame,
+    columns: List[str],
+) -> list:
+    """Find all string values occurring in given columns of a DataFrame"""
+    # limit our analysis to columns that contain strings
+    # (or other object types)
+    cols_with_strs = (
+        data[columns].select_dtypes(include=[object]).columns.values.tolist()
+    )
+    temp = []
+    for col in cols_with_strs:
+        temp += list(data[col].unique())
+    temp = list(set(temp))
+    strs = [x for x in temp if not is_float(x)]
+    return strs
+
+
+def parse_code(code: str) -> float:
+    """Parse a string code and return 0 or np.nan based on rules to interpret
+    the codes."""
+    code = code.strip()
+    if code in _special_codes:
+        return _special_codes[code]
+
+    parts = code.split(",")
+    parts = [x.replace(".", "").strip() for x in parts]
+    if "IE" in parts or "NO" in parts:
+        return 0
+    if "NE" in parts or "NA" in parts:
+        return np.nan
+    raise ValueError(f"Could not parse code: {code!r}.")
+
+
+def create_na_replacement_dict(
+    strs: List[str],
+    user_na_conv: Dict[str, str] = {},
+) -> Dict[str, str]:
+    """Create a dict for replacement of strings by NaN and 0 based on
+    general rules and user defined rules"""
+
+    mapping = {}
+    for str_val in strs:
+        if str_val in user_na_conv:
+            mapping[str_val] = user_na_conv[str_val]
+        else:
+            mapping[str_val] = parse_code(str_val)
+
+    return mapping
+
+
+def replace_na_values(data: pd.DataFrame, columns: List[str], na_repl_dict):
+    """Replace str values indicating not-a-number by float NaN."""
```

Comment:
It doesn't strictly use NA values, right? Default rules also have 0, and user rules could map stuff to 1.5 if they wanted.

Reply:
So maybe `replace_values` would be a better name to avoid confusion?

Reply:
right